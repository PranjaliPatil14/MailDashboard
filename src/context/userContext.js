import React, { useEffect, useState } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { USER } from "../utilities/useLocalStorage";

const initialUserContextValue = {
  users: [],
  loggedInUser: {
    id: 0,
    name: "",
    role: "",
    profileImage: "",
  },
  setLoggedInUser: () => {},
  setUsers: () => {},
};

export const UserContext = React.createContext(initialUserContextValue);

const UserContextProvider = ({ children }) => {
  const [users, setUsers] = useState([]);
  const [loggedInUser, setLoggedInUser] = useState({});
  const { value } = useLocalStorage(USER);

  useEffect(() => {
    if (!users.length) {
      setUsers(value);
    }
  }, [users.length, value]);

  return (
    <UserContext.Provider
      value={{ users, setUsers, loggedInUser, setLoggedInUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

UserContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default UserContextProvider;
