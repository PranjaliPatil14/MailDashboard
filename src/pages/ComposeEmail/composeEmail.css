@import "../../assets/styles/colors.css";

.compose-mail-form {
    position: absolute;
    bottom: 0;
    right: 0;
    background: white;
    display: flex;
    flex-direction: column;
    margin: 30px;
    width: 700px;
     color: #686C6E;
}

.recipients {
    display: flex;
    flex-direction: column;
}

.recipients > input{
   padding: 20px;
}

.compose-mail--body > textarea{
    padding: 20px;
    min-height: 300px;
    resize: none;
    width: 100%;
    margin: 0;
    box-sizing: border-box;
}

.compose-mail--body {
    flex: 1;
}

.send-mail {
    background: var(--green);
    padding: 15px 30px;
    color: white;
    font-size: 16px;
    font-weight: bold;
    border-radius: 5px;
}
