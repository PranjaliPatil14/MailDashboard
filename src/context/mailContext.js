import React, { useContext, useEffect, useReducer } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { MAIL } from "../utilities/useLocalStorage";
import { UserContext } from "./userContext";
import {
  mailReducer,
  SET_LOGGED_IN_USER_MAIL,
  SET_VALUES,
} from "./mailReducer";

const initialMailContextValue = {
  mails: [],
  refresh: true,
  loggedInUserMails: [],
  dispatch: () => {},
};

export const MailContext = React.createContext(initialMailContextValue);

const MailContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(mailReducer, initialMailContextValue);
  const { value } = useLocalStorage(MAIL);
  const { loggedInUser } = useContext(UserContext);

  useEffect(() => {
    if (state.refresh) {
      dispatch({ type: SET_VALUES, data: value });
    }
  }, [value, state.refresh]);

  useEffect(() => {
    dispatch({ type: SET_LOGGED_IN_USER_MAIL, data: loggedInUser.id });
  }, [loggedInUser, value, state.mails]);

  useEffect(() => {
    localStorage.setItem(MAIL, JSON.stringify(state.mails));
  }, [state.mails]);

  return (
    <MailContext.Provider value={{ ...state, dispatch }}>
      {children}
    </MailContext.Provider>
  );
};

MailContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default MailContextProvider;
