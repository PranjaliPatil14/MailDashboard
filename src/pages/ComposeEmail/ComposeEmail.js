import React, { useContext, useState } from "react";
import * as PropTypes from "prop-types";
import Button from "../../components/button/Button";
import { UserContext } from "../../context/userContext";
import "./composeEmail.css";
import { MailContext } from "../../context/mailContext";
import { SEND_MAIL } from "../../context/mailReducer";

const ComposeEmail = ({ setShowCompose }) => {
  const [to, setTo] = useState("");
  const [cc, setCC] = useState("");
  const [body, setBody] = useState("");

  const { loggedInUser, users } = useContext(UserContext);
  const { dispatch } = useContext(MailContext);

  const onSubmit = () => {
    const toMail = users.find((user) => user.email === to).id;
    const ccMail = users.find((user) => user.email === cc)?.id ?? -1;
    const newMail = {
      id: `${Math.random() * 1000}-${new Date().toString()}`,
      from: loggedInUser.id,
      to: [toMail],
      cc: [ccMail],
      body,
      attachment: false,
      time: new Date().toString(),
      tag: "",
      read: false,
      important: false,
    };
    dispatch({ type: SEND_MAIL, data: newMail });
    setTimeout(() => setShowCompose(false));
  };
  return (
    <div className="compose-mail-form">
      <div className="recipients">
        <input
          onChange={({ target }) => setTo(target.value)}
          value={to}
          placeholder="TO"
          data-testid="to"
        />
        <input
          onChange={({ target }) => setCC(target.value)}
          value={cc}
          placeholder="CC"
          data-testid="cc"
        />
      </div>
      <div className="compose-mail--body">
        <textarea
          onChange={({ target }) => setBody(target.value)}
          value={body}
          data-testid="body"
          placeholder="Please write your message here"
        />
      </div>
      <Button onClick={onSubmit} className="send-mail">
        Send
      </Button>
    </div>
  );
};

ComposeEmail.propTypes = {
  setShowCompose: PropTypes.func.isRequired,
};

export default ComposeEmail;
