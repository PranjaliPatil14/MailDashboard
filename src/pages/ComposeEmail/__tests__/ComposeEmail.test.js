import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ComposeEmail from "../ComposeEmail";
import { MailContext } from "../../../context/mailContext";
import { UserContext } from "../../../context/userContext";
import mockUsers from "../../../mockData/users.json";

describe("composeEmail", () => {
  const setShowCompose = jest.fn();
  const dispatch = jest.fn();
  const loggedInUser = mockUsers[0];
  const users = mockUsers;
  let component;
  beforeEach(() => {
    component = render(
      <UserContext.Provider value={{ loggedInUser, users }}>
        <MailContext.Provider value={{ dispatch }}>
          <ComposeEmail setShowCompose={setShowCompose} />
        </MailContext.Provider>
      </UserContext.Provider>
    );
  });
  test("should render compose email form", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

  test("should call dispatch with send mail action", () => {
    jest.useFakeTimers();
    const mockDate = new Date(1466424490000);
    const spy = jest.spyOn(global, "Date").mockImplementation(() => mockDate);
    global.Math.random = () => 0.5;

    fireEvent.change(component.getByTestId("to"), {
      target: { value: "pqr@gmail.com" },
    });
    fireEvent.change(component.getByTestId("cc"), {
      target: { value: "mat@gmail.com" },
    });
    fireEvent.change(component.getByTestId("body"), {
      target: { value: "hello" },
    });
    fireEvent.click(component.getByText("Send"));

    expect(dispatch).toHaveBeenCalledWith({
      data: {
        attachment: false,
        body: "hello",
        cc: [3],
        from: 1,
        id: "500-Mon Jun 20 2016 17:38:10 GMT+0530 (India Standard Time)",
        important: false,
        read: false,
        tag: "",
        time: "Mon Jun 20 2016 17:38:10 GMT+0530 (India Standard Time)",
        to: [2],
      },
      type: "SEND_MAIL",
    });

    jest.advanceTimersByTime(1001);
    expect(setShowCompose).toHaveBeenCalledWith(false);
    spy.mockRestore();

    jest.useRealTimers();
  });

  test("should call send mail with only to receipient", () => {
    jest.useFakeTimers();
    const mockDate = new Date(1466424490000);
    const spy = jest.spyOn(global, "Date").mockImplementation(() => mockDate);
    global.Math.random = () => 0.5;

    fireEvent.change(component.getByTestId("to"), {
      target: { value: "pqr@gmail.com" },
    });

    fireEvent.change(component.getByTestId("body"), {
      target: { value: "hello" },
    });
    fireEvent.click(component.getByText("Send"));

    expect(dispatch).toHaveBeenCalledWith({
      data: {
        attachment: false,
        body: "hello",
        cc: [-1],
        from: 1,
        id: "500-Mon Jun 20 2016 17:38:10 GMT+0530 (India Standard Time)",
        important: false,
        read: false,
        tag: "",
        time: "Mon Jun 20 2016 17:38:10 GMT+0530 (India Standard Time)",
        to: [2],
      },
      type: "SEND_MAIL",
    });

    jest.advanceTimersByTime(1001);
    expect(setShowCompose).toHaveBeenCalledWith(false);
    spy.mockRestore();

    jest.useRealTimers();
  });
});
