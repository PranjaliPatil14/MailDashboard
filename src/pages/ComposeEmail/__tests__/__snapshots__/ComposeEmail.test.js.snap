// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeEmail should render compose email form 1`] = `
<DocumentFragment>
  <div
    class="compose-mail-form"
  >
    <div
      class="recipients"
    >
      <input
        data-testid="to"
        placeholder="TO"
        value=""
      />
      <input
        data-testid="cc"
        placeholder="CC"
        value=""
      />
    </div>
    <div
      class="compose-mail--body"
    >
      <textarea
        data-testid="body"
        placeholder="Please write your message here"
      />
    </div>
    <button
      class="button send-mail"
      data-testid=""
      type="submit"
    >
      Send
    </button>
  </div>
</DocumentFragment>
`;
